### DISCUSSION TITLE
can ghidra do  dynamic analysis  thing ?

### DISCUSSION THREAD

**Discussion by cattrace** [NationalSecurityAgency/ghidra#24] [Debugger]
can ghidra do  dynamic analysis  thing ?

**Comment by CharlesDardaman** [comment]
If you are asking if it has a debugger the answer is not at this time.

**Comment by eacmen** [comment]
![image](https://user-images.githubusercontent.com/21159128/53910165-33e11c00-4021-11e9-968a-55fd406ccd8c.png)

**Comment by nsadeveloper789** [comment]
It's one of the in-development features announced at RSA.

**Comment by gettysburg** [comment]
Any ETA on the debugger?

**Comment by nsadeveloper789** [comment]
We're still working on it. It's coming soon, but we don't have an ETA, yet.

**Comment by efeozyer** [comment]
any news with debugger?

**Comment by nsadeveloper789** [comment]
Still in the works.

**Comment by TheAifam5** [comment]
Can we know the progress of the debugger, like what's done and what's in progress? It's been like 1 year after the announcement at RSA.

**Comment by nsadeveloper789** [comment]
We're in the process of internal testing. Depending on how that goes, we'd like to merge it sometime in the 9.2.x series. Honestly, we're probably as excited/anxious to get it out there as you are to see it! As it stands, the alpha debugging framework can connect with WinDbg or GDB; display registers, memory listings, etc.; incorporate static analysis; record/replay sessions in the project database; etc. As we approach release, please understand it has taken us some time to get this code into a stable condition, and many areas are still incubating.

**Comment by TheAifam5** [comment]
@nsadeveloper789 Thank your for your response. I appreciate all the work on it and will be patient for the release of the debugger.

**Comment by Bue-von-hon** [comment]
> We're in the process of internal testing. Depending on how that goes, we'd like to merge it sometime in the 9.2.x series. Honestly, we're probably as excited/anxious to get it out there as you are to see it! As it stands, the alpha debugging framework can connect with WinDbg or GDB; display registers, memory listings, etc.; incorporate static analysis; record/replay sessions in the project database; etc. As we approach release, please understand it has taken us some time to get this code into a stable condition, and many areas are still incubating.

thank you for working

**Comment by Eugenio-Bruno** [comment]
Hi! I'm going to be "that guy" and ask if there's been any progress?

My crystal ball* says ghidra 9.2 with debugging support will be released in the next few months.
*Please note my crystal ball works entirely on hopes and dreams

**Comment by gettysburg** [comment]
> 
> 
> Hello. Any news on this? Have been putting off learning more exploit development stuff until this debugger is released. A timeframe would be nice.

There's tons of other debuggers you can use, including IDA's debugger which is one of the best in my opinion but I can only comment on version 7.2 until I get a proper license. If you are not a pirate, x64dbg is decent, my friend uses it all the time and so do I, occasionally - has all the features you need, decent UI, very fast, decent plugin API plus an active and friendly community.

One of the developers already mentioned a time frame - remember, this is free software, give them a break, they are not getting paid to implement features for you.

**Comment by vn-ki** [comment]
Can we please keep this thread civil and on topic?

I think I'm talking for many when I say that people subscribed to this issue doesn't care about your personal feuds. 

Please take it somewhere else.

**Comment by ryanmkurtz** [comment]
I can no longer lock issues, but I put a request in.  Please consider it locked and stop commenting.

As for the original question: can Ghidra debug?  I think we all know that it currently can't, they are working very hard on it, COVID has slowed things down, and when it is ready, they will announce it.

I think this issue/question can be closed.

**Comment by emteere** [comment]
As @ryanmkurtz mentioned work is continuing on the debugger, it is a large job and there have been many issues that have come up, not the least of which is COVID.  We would really like to give you a timeline, but I don't want to promise anything and not deliver in a timely manner.  Dynamic analysis and easy integration of dynamic information is important to the Ghidra community as is evidenced in the thread.

IMHO, I would not hesitate to learn/use any tool, as the knowledge gained can always be ported to another tool.  Time spent learning other debuggers, such as gdb or windbg, will not be wasted as the new Ghidra based debugger will be integrated with existing debuggers or have considerable overlap with their command sets and functionality. That said, if another tool fits your needs, use it.  There are many great things about each RE tool and reasons to use or not use each.

Rest assured we are diligently working on the debugger and we think it will be worth the wait.  We'll share information on features and release when we have solid timelines.

Thanks for using Ghidra in whatever way fits your work.  Feel free to request features/improvements you see lacking and we and the community will try to address them.

We'll close the issue for now.

**Comment by eacmen** [comment]
https://twitter.com/jasongeffner/status/1287197839144165378?s=21

**Comment by ghost** [comment]
nsadeveloper - random.randint(0,999)

**Comment by nsadeveloper789** [comment]
Why is 6 so afraid of nsadeveloper7?

**Comment by justanotheranonymoususer** [comment]
Because @nsadeveloper789!

![image](https://user-images.githubusercontent.com/5781692/95077384-148a0380-071c-11eb-9d35-2b2eda6bf223.png)

**Comment by justanotheranonymoususer** [comment]
For those who missed 🎉

> Today our **#Ghidra** team pushed a preview debugger on 
> @GitHub. **#Developers** and testers, check out this long awaited feature:
> https://github.com/NationalSecurityAgency/ghidra/tree/debugger

https://twitter.com/NSACyber/status/1339652646513291264

**Comment by Wall-AF** [comment]
(How) can it be used with Windows 3.11???

**Comment by nsadeveloper789** [comment]
This initial preview release is geared primarily toward modern x86_64 applications on Windows and Linux. That said, the framework consists of many re-usable components, and integrates with existing debuggers, allowing you to try piecing together just about anything you'd like, but you may need to do a fair bit of coding.

If you're serious, try Ghidra's GDB agent, connected to qemu, running a Windows 3.11 VM. Chances are, the connector and "Objects" viewer will work out-of-the-box, but displaying machine state, correlating it to imported programs, and other sensemaking, will at the very least require you to write a mapper/opinion.

**Comment by SirEnder125** [comment]
Does Ghidra 9.2.2 have a gdb debugger tool?

**Reply by ghidra1** [reply]
No.  The Ghidra debugger tool is in development and is not contained with 9.2.x

You can currently access the debugger via its development branch on GitHub.

**Reply by SirEnder125** [reply]
Uhh I actually just downloaded and unzipped that and it says JDK was not found...
But v.9.2.2 works fine??

**Reply by ghidra1** [reply]
You will need to follow the [Developer Guide](https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md) for setting up an Eclipse environment from which to run Ghidra from the `debugger` branch.  Running from an installation of Ghidra (e.g., 9.2.2) utilizes scripts which attempt to identify your Java installation - this does not occur when running in a development setup.

**Reply by SirEnder125** [reply]
Oh, okay. It is different from 9.2.2, hm? I did not need to do that with that version. I'll try it. THanks.

**Reply by SirEnder125** [reply]
Tell me, will the release version of the debugger run automatically? Without eclipse?

**Reply by ryanmkurtz** [reply]
There are 2 sections of the Developer's Guide you should look at:

- Building Ghidra
- Developing Ghidra

If you have the debugger branch checked out in git, the Building Ghidra section will produce a custom build of Ghidra that is similar to the releases you find on ghidra-sre.org.  You will need gradle to build Ghidra, but not Eclipse.

Eclipse comes into play when you are following the Developing Ghidra instructions.  If all you care about is testing out the debugger, then you don't need Eclipse.  Just follow the instructions up to and including "Building Ghidra."

**Reply by ghidra1** [reply]
> Tell me, will the release version of the debugger run automatically? Without eclipse?

When it is merged into the master branch it will be incorporated into in the subsequent major release of Ghidra and would not require Eclipse to be used.

**Comment by SirEnder125** [comment]
Soooo, when is 9.3 (debugger release) coming out? (Just a close guess is fine)

**Reply by coldwaterq** [reply]
I'm trying to decide if I should figure out how to build the latest version or if I can just wait for it to be released in a couple weeks, so even just an ETA would be super helpful.

**Reply by TheAifam5** [reply]
Building ghidra is pretty straightforward ;) The process described there looks pretty messy but at the end you need to execute only few commands (clone, gradle)

**Reply by emteere** [reply]
We can't give you definitive date yet (approximately a month or two).  The debugger will be in our next release since it has been merged into the master branch.

As @TheAifam5 mentions, the build process is fairly straightforward, and is even more so in the current master branch.  We've pre-built some parsers that required lex/yacc(flex/bison) so that you don't need anything more than JDK 11 (which you would need anyway), a compiler for your platform, and gradle.

**Reply by emteere** [reply]
We need to update the DevGuide readme, but the current minimal build on windows would be:

Download a .zip clone of ghidra repository
Installed VisualStudio
Installed Gradle
Installed JDK-11
Run gradle --init
Run gradle buildGhidra

built DEV windows distribution will be in ghidra/build/dist
